src/core/Config.cpp

```diff
@@ -337,44 +337,55 @@ void Config::setDefaultActionKeys() {
 	}
 }
 
-void Config::setActionKey(ControlAction actionId, int index, InputKeyId key) {
+void Config::setActionKey(const ControlAction actionId, uint32_t index, InputKeyId newKey) {
 	
-	if(actionId < 0 || (size_t)actionId >= NUM_ACTION_KEY || index > 1 || index < 0) {
+	if(actionId < 0 || (size_t)actionId >= NUM_ACTION_KEY || index >= NUM_OF_KEYS_PER_ACTION) {
 		arx_assert(false);
 		return;
 	}
 	
 	ActionKey & action = actions[actionId];
-	
 	InputKeyId oldKey = action.key[index];
-	action.key[index] = key;
-	
-	int otherIndex = 1 - index;
-	
-	if(action.key[otherIndex] == -1) {
-		action.key[otherIndex] = oldKey;
-		oldKey = -1;
-	}
+	action.key[index] = newKey;
 	
-	if(action.key[otherIndex] == key) {
-		action.key[otherIndex] = -1;
-	}
+	uint32_t otherIndex = ALT_ACTION - index;
 	
-	// remove double key assignments
-	for(size_t i = 0; i < NUM_ACTION_KEY; i++) {
-		
-		if(i == (size_t)actionId) {
-			continue;
-		}
-		
-		for(int k = 0; k < 2; k++) {
-			if(actions[i].key[k] == key) {
-				actions[i].key[k] = oldKey;
-				oldKey = -1;
-			}
-		}
-		
-	}
+	moveOldKeyToOtherIndexIfEmpty(otherIndex, oldKey, action);
+	clearOtherIndexIfDuplicate(otherIndex, newKey, action);
+	removeDuplicateKeyAssignments(actionId, newKey, oldKey);
+}
+
+void Config::moveOldKeyToOtherIndexIfEmpty(uint32_t otherIndex,
```

Comment:
It really depends on how small they and on the name & arguments. The point is to make the code as readable as possible not to have some absolute rules about using functions instead of comments.

Some concrete points:
* For the first two functions, the code is so short it its obvious what it does and doesn't need a comment. Moving it into a function only requires the reader to jump around in the file to follow the code flow and understand the overall bigger function.
* Your extracted function names describe what the functions are used for instead of what they do: `moveOldKeyToOtherIndexIfEmpty` does not care that the key is old, `moveOldKeyToOtherIndexIfEmpty` and `clearOtherIndexIfDuplicate` don't care about the index not being the primary one - they only deal with one index, and `removeDuplicateKeyAssignments` does not care about duplicates at all - it simply replaces the first use of a key with a supplied value and all other uses (there should be none if the bindings already are consistent!) with `UNASSIGNED_KEY`.
* The names `moveOldKeyToOtherIndexIfEmpty` and `clearOtherIndexIfDuplicate` do not clearly describe what is checked for being empty/duplicate. The key or index aren't empty, the binding slot is. The index isn't duplicated, the key binding is.
* Having conditions like "if", "except", etc. in your function names is a good indicator that your abstraction is not at the right level.

A *good* way to extract a function here could be:
```cpp
/*!
 * \brief Bind a new key to the action
 *
 * \param index Index of the slot to bind the key to.
 *
 * \return the previously bound key if it could not be moved to the other slow slot,
 *         or \ref UNASSIGNED_KEY otherwise
 */
InputKeyId ActionKey::changeBinding(int index, InputKeyId newKey) {
	int otherIndex = 1 - index;
	InputKeyId oldKey = key[index];
	key[index] = newKey;
	if(key[otherIndex] == UNASSIGNED_KEY) {
		key[otherIndex] = oldKey;
		oldKey = UNASSIGNED_KEY;
	}
	if(key[otherIndex] == newKey) {
		key[otherIndex] = UNASSIGNED_KEY;
	}
	return oldKey;
}
```
Possibly made private or not as a member function though to discourage calling it directly.
That way:
-  The concerns are neatly separated: The extracted function only cares about consistency in one `ActionKey` while the larger `Config::setActionKey` does not need to care about there being multiple bindings per action.
- The function conceptually only modifies one thing (the ActionKey) as opposed to two things in your `moveOldKeyToOtherIndexIfEmpty` (the ActionKey binding at otherIndex and the oldKey value).
- Returning the old value when changing something is a common idiom. The fact that there are two slots and you are only returning something if both slots are full is an implementation detail.
- Using it as `InputKeyId removedKey = action.changeBinding(index, key)` would neatly document what you are doing.

`removeDuplicateKeyAssignments` has a bit more reason for existing as a separate function. However, the code is still not very long, `Config::setActionKey` does not contain much else after `ActionKey::changeBinding` has been extracted, and "replace all uses of this key except for those in this action" does not make for a nice function name. An argument *for* extracting it could be that there is potential to reuse it to sanitize bindings while loading the config.

Hope this helps understand my reasoning and I hope I'm not coming off as too discouraging ;-)